flow/Trace.h

```diff
@@ -135,20 +135,32 @@ template <class T> class Standalone;
 template <class T> class Optional;
 
 struct TraceEvent {
-	TraceEvent( const char* type, UID id = UID() );   // Assumes SevInfo severity
-	TraceEvent( Severity, const char* type, UID id = UID() );
+	// Assumes SevInfo severity
+	TraceEvent( const char* type, UID id = UID(), double suppressFor = 0.0 );
+	TraceEvent( const char* type, double suppressFor );
+	TraceEvent( const char* type, const StringRef& id);
+	TraceEvent( const char* type, const Optional<Standalone<StringRef>>& id);
+	TraceEvent( const char* type, const class Error& e, double suppressFor = 0.0 );
```

Comment:
Is there something better we can call `e` here? If I understand correctly, it's used only to disable the trace event in the case that it's an `actor_cancelled` error, but that's not clear from this signature. To be honest, I think the idea of passing in an error meant to disable the trace event in a specific case may lead to some confusion for people even if the signature is made more explicit, so it might be worth taking a second trying to imagine some other pattern that feels more natural. It's likely that an alternate solution would require deferring any substantive work until after the constructor (e.g. wait until first detail call or destructor, whichever is first), and it's possible we could look to other logging implementations for inspiration. If there's nothing else that seems reasonable, though, then we should do our best to make this as clear as possible.

Reply:
Passing an error to the constructor also logs the error. Previously, there were two uses of .error. 90% of the use cases wanted to log an error and suppress the trace event if the message was actor_cancelled. These use cases all now pass the error into the constructor. This is equivalent to passing in an ID to the constructor instead of calling .detail("ID",id) which would also do the same thing.

The other use case for .error is that you want to see actor cancelled messages. This is now accomplished by passing the error the the .error function, which is no longer conditional and now will always log errors.

To me, especially because of the symmetry with passing ID into the constructor, it feels natural to pass an error into the constructor.

Reply:
Ah, I was expecting to see it logged in init, but I guess you aren't doing that because init isn't taking an optional error or doing something else to distinguish the cases where an error isn't passed in. 

I think I might like to see that behavior moved to init because we've previously run into issues with some of the constructors not being kept up to date properly.

I agree this behavior with the error parameter is better, but I would still personally be in favor of giving a little thought to whether there's a reasonable API that isn't based on passing a bunch of optional parameters to a constructor.

Reply:
I suppose if you didn't want to move the error logging in so you could keep the ordering with ID the same, that could be reasonable. Maybe in that case you could not pass the error to init and instead pass the specific info you want (whether the event should be suppressed).

Reply:
I implemented your suggestion with init. After doing the conversion I am pretty happy with how the trace events look now, but I will solicit office opinions on the topic.